File: backend/internal/handler/dto/playground.go
package dto

import (
	"encoding/json"
	"time"

	"github.com/Wei-Shaw/sub2api/internal/service"
)

// PlaygroundChatRequest 对话广场聊天请求的关键字段提取结构。
// 仅用于参数校验与中间件分发，其余 OpenAI 标准字段（temperature/top_p/...）
// 由上游 ChatCompletions handler 通过 gjson 直接读取请求体处理，无需在此声明。
//
// 注意：本结构与 OpenAI ChatCompletion 请求体共存于同一 JSON，
// 反序列化时未声明字段会被忽略，原始请求体仍可完整透传到上游。
type PlaygroundChatRequest struct {
	// Model 用户选择的模型 ID（必填）
	Model string `json:"model"`
	// Group 用户选择的分组名（必填，需属于用户可用分组列表）
	Group string `json:"group"`
}

// PlaygroundAvailableModel 对话广场可用模型条目
type PlaygroundAvailableModel struct {
	// ID 模型唯一标识（如 "claude-sonnet-4-20250514"）
	ID string `json:"id"`
	// Platform 模型所属平台（"openai" / "anthropic" / "gemini" / "antigravity"）
	Platform string `json:"platform"`
	// GroupID 模型来源分组 ID
	GroupID int64 `json:"group_id"`
	// GroupName 模型来源分组名
	GroupName string `json:"group_name"`
}

// PlaygroundAvailableModelsResponse 可用模型列表响应
type PlaygroundAvailableModelsResponse struct {
	// Models 当前查询分组下用户可用的模型列表
	Models []PlaygroundAvailableModel `json:"models"`
}

// ---- 会话 CRUD DTO ----

// PlaygroundConversationSummaryDTO 会话列表条目（不含 messages 大字段）。
// 用于 GET /conversations 列表接口。
type PlaygroundConversationSummaryDTO struct {
	// ID 会话主键
	ID int64 `json:"id"`
	// Title 会话标题
	Title string `json:"title"`
	// Model 最后使用的模型标识（可选）
	Model *string `json:"model,omitempty"`
	// GroupName 所属分组名称（可选）
	GroupName *string `json:"group_name,omitempty"`
	// LastActivityAt 最后活动时间
	LastActivityAt time.Time `json:"last_activity_at"`
	// CreatedAt 创建时间
	CreatedAt time.Time `json:"created_at"`
}

// PlaygroundConversationDetailDTO 会话详情（含 messages）。
// 用于 GET /conversations/:id 接口。
type PlaygroundConversationDetailDTO struct {
	// ID 会话主键
	ID int64 `json:"id"`
	// Title 会话标题
	Title string `json:"title"`
	// Model 最后使用的模型标识（可选）
	Model *string `json:"model,omitempty"`
	// GroupName 所属分组名称（可选）
	GroupName *string `json:"group_name,omitempty"`
	// Messages 完整消息列表，JSONB 原样透传
	Messages json.RawMessage `json:"messages"`
	// LastActivityAt 最后活动时间
	LastActivityAt time.Time `json:"last_activity_at"`
	// CreatedAt 创建时间
	CreatedAt time.Time `json:"created_at"`
	// UpdatedAt 更新时间
	UpdatedAt time.Time `json:"updated_at"`
}

// CreateConversationRequest 新建会话请求体。
type CreateConversationRequest struct {
	// Title 会话标题（必填）
	Title string `json:"title"`
	// Model 使用的模型 ID（可选）
	Model *string `json:"model,omitempty"`
	// GroupName 所属分组名称（可选）
	GroupName *string `json:"group_name,omitempty"`
	// Messages 初始消息列表（可选，JSONB 原样存储）
	Messages json.RawMessage `json:"messages,omitempty"`
}

// UpdateConversationRequest 更新会话请求体，所有字段均可选（nil 表示不改）。
// model/group_name 传 null 表示清空，不传（字段缺失）与传 null 语义相同。
type UpdateConversationRequest struct {
	// Title 会话标题（可选，传入则更新）
	Title *string `json:"title,omitempty"`
	// Model 模型标识（可选，传入则更新，传 null 则清空）
	Model *string `json:"model"`
	// GroupName 分组名称（可选，传入则更新，传 null 则清空）
	GroupName *string `json:"group_name"`
	// Messages 完整消息列表（可选，传入则覆盖更新）
	Messages json.RawMessage `json:"messages,omitempty"`
}

// ConversationSummaryFromService 将 service 摘要结构转换为 DTO。
func ConversationSummaryFromService(s *service.PlaygroundConversationSummary) *PlaygroundConversationSummaryDTO {
	if s == nil {
		return nil
	}
	return &PlaygroundConversationSummaryDTO{
		ID:             s.ID,
		Title:          s.Title,
		Model:          s.Model,
		GroupName:      s.GroupName,
		LastActivityAt: s.LastActivityAt,
		CreatedAt:      s.CreatedAt,
	}
}

// ConversationDetailFromService 将 service 完整会话结构转换为 DTO。
func ConversationDetailFromService(c *service.PlaygroundConversation) *PlaygroundConversationDetailDTO {
	if c == nil {
		return nil
	}
	return &PlaygroundConversationDetailDTO{
		ID:             c.ID,
		Title:          c.Title,
		Model:          c.Model,
		GroupName:      c.GroupName,
		Messages:       c.Messages,
		LastActivityAt: c.LastActivityAt,
		CreatedAt:      c.CreatedAt,
		UpdatedAt:      c.UpdatedAt,
	}
}

File: frontend/src/types/playground.ts
/**
 * 对话广场（Playground）类型定义
 * 移植自 new-api/web/default/src/features/playground/types.ts
 * 适配 sub2api Vue 3 技术栈
 */

// ==================== 消息相关 ====================

export type MessageRole = 'user' | 'assistant' | 'system'

export type MessageStatus = 'loading' | 'streaming' | 'complete' | 'error'

export interface MessageVersion {
  /** 版本唯一 ID（用于多版本切换的 key） */
  id: string
  /** 该版本对应的文本内容 */
  content: string
  /** 视频状态跟随版本保存，避免重新生成后结果串位。 */
  video?: PlaygroundVideoState
}

export type PlaygroundVideoStatus =
  | 'creating'
  | 'queued'
  | 'in_progress'
  | 'completed'
  | 'failed'
  | 'stopped'

export interface PlaygroundVideoState {
  id?: string
  status: PlaygroundVideoStatus
  progress: number
  url?: string
}

export interface MessageReasoning {
  /** 推理内容（思考链） */
  content: string
  /** 推理耗时（毫秒） */
  duration: number
}

export interface MessageSource {
  href: string
  title: string
}

export type PlaygroundAttachmentKind = 'image' | 'document'

export interface PlaygroundAttachment {
  id: string
  kind: PlaygroundAttachmentKind
  name: string
  type: string
  size: number
  /** 图片使用 data URL 直传给 OpenAI 兼容 image_url。 */
  dataUrl?: string
  /** 文档在浏览器本地读取为纯文本后拼入用户消息。 */
  text?: string
}

export interface Message {
  /** 消息唯一 key */
  key: string
  /** 消息来源（user/assistant/system） */
  from: MessageRole
  /** 多版本数组：versions[0] 为当前显示版本；regenerate 时追加新版本 */
  versions: MessageVersion[]
  /** 引用来源列表 */
  sources?: MessageSource[]
  /** 推理内容（reasoning_content / <think> 标签解析） */
  reasoning?: MessageReasoning
  /** 推理是否在流式输出中 */
  isReasoningStreaming?: boolean
  /** 推理是否已完成 */
  isReasoningComplete?: boolean
  /** 内容是否已完成 */
  isContentComplete?: boolean
  /** 当前状态 */
  status?: MessageStatus
  /** 错误码（用于本地化错误提示） */
  errorCode?: string | null
  /** 用户上传的图片/文本类文档附件 */
  attachments?: PlaygroundAttachment[]
}

// ==================== API 请求/响应 ====================

export interface ContentPart {
  type: 'text' | 'image_url'
  text?: string
  image_url?: { url: string }
}

export interface ChatCompletionMessage {
  role: MessageRole
  content: string | ContentPart[]
}

export interface ChatCompletionRequest {
  model: string
  group: string
  messages: ChatCompletionMessage[]
  stream: boolean
  temperature?: number
  top_p?: number
  max_tokens?: number
  frequency_penalty?: number
  presence_penalty?: number
  seed?: number | null
}

export interface ChatCompletionChunkDelta {
  role?: MessageRole
  content?: string
  reasoning_content?: string
}

export interface ChatCompletionChunk {
  id: string
  object: string
  created: number
  model: string
  choices: Array<{
    index: number
    delta: ChatCompletionChunkDelta
    finish_reason: string | null
  }>
}

export interface ChatCompletionResponse {
  id: string
  object: string
  created: number
  model: string
  choices: Array<{
    index: number
    message: {
      role: MessageRole
      content: string
      reasoning_content?: string
    }
    finish_reason: string
  }>
  usage?: {
    prompt_tokens: number
    completion_tokens: number
    total_tokens: number
  }
}

export interface PlaygroundVideoRequest {
  model: string
  group: string
  prompt: string
  seconds?: string
  aspect_ratio?: string
  input_reference?: { image_url: string }
}

export interface PlaygroundVideoResponse {
  id: string
  model?: string
  object?: string
  progress?: number | string
  seconds?: number | string
  status: string
  video_url?: string | null
  video?: { url?: string }
  error?: string | { message?: string; error?: string; code?: string }
  message?: string
}

export interface PlaygroundImageRequest {
  model: string
  group: string
  prompt: string
  n?: number
  size: string
  image?: string | string[] | Record<string, unknown>
  quality?: string
  response_format?: string
  style?: string
  background?: string
  watermark?: boolean
}

export interface PlaygroundImageResponse {
  created?: number
  data?: Array<{
    url?: string
    b64_json?: string
    revised_prompt?: string
  }>
  error?: string | { message?: string; error?: string; code?: string }
  message?: string
}

// ==================== 配置 ====================

export interface PlaygroundConfig {
  model: string
  group: string
  videoSeconds: string
  videoAspectRatio: string
  imageSize: string
  imageQuality: string
  imageResponseFormat: string
  imageStyle: string
  imageBackground: string
  imageWatermark: boolean
  temperature: number
  top_p: number
  max_tokens: number
  frequency_penalty: number
  presence_penalty: number
  seed: number | null
  stream: boolean
  systemPrompt: string
}

export interface ParameterEnabled {
  temperature: boolean
  top_p: boolean
  max_tokens: boolean
  frequency_penalty: boolean
  presence_penalty: boolean
  seed: boolean
  imageQuality: boolean
  imageResponseFormat: boolean
  imageStyle: boolean
  imageBackground: boolean
  imageWatermark: boolean
}

// ==================== 模型/分组 ====================

export interface ModelOption {
  /** 显示文本 */
  label: string
  /** 实际 model ID */
  value: string
  /** 所属平台（用于按平台禁用某些参数） */
  platform?: string
}

export interface GroupOption {
  /** 显示文本 */
  label: string
  /** 实际 group name */
  value: string
  /** 分组费率倍数 */
  ratio: number
  /** 平台（openai / anthropic / gemini / antigravity） */
  platform?: string
  /** 简介 */
  desc?: string
}

// ==================== 后端响应 ====================

export interface PlaygroundAvailableModel {
  id: string
  platform: string
  group_id: number
  group_name: string
}

export interface PlaygroundAvailableModelsResponse {
  models: PlaygroundAvailableModel[]
}

// ==================== 多会话持久化 ====================

/** 会话摘要（列表条目，不含 messages 大字段） */
export interface ConversationSummary {
  id: number
  title: string
  model?: string | null
  group_name?: string | null
  /** 最后活动时间（RFC3339），列表按此倒序 */
  last_activity_at: string
  created_at: string
}

/** 会话详情（含完整消息列表，JSONB 原样透传） */
export interface ConversationDetail extends ConversationSummary {
  /** 消息列表；新建空会话时后端可能返回 null */
  messages: Message[] | null
  updated_at: string
}

/** 新建会话请求体 */
export interface CreateConversationRequest {
  title: string
  model?: string | null
  group_name?: string | null
  messages?: Message[]
}

/**
 * 更新会话请求体。
 *
 * ⚠️ 后端语义：title/messages 缺省 = 不改；model/group_name 缺省或 null = 清空！
 * 因此每次保存消息时必须同时带上 model 和 group_name。
 */
export interface UpdateConversationRequest {
  title?: string
  model: string | null
  group_name: string | null
  messages?: Message[]
}

// ==================== 错误类型 ====================

export interface PlaygroundError {
  type: string
  message: string
  code?: string
}
